Bluetooth: Mesh: Add full virtual addresses support

PavelVPV opened the pull request:
This PR adds the following features related to virtual addresses
support:
- Allows to store Label UUIDs which virtual addresses collide;
- Allows to decrypt messages encrypted with a virtual address with
collision;
- Allows to publish a message to a specific Label UUID to avoid virtual
addresses collision by adding a pointer to Label UUID to
struct bt_mesh_msg_ctx and struct bt_mesh_model_pub;
- Allows to differentiate Label UUIDs in the model's Subscription List
by storing all subscribed UUIDs in struct bt_mesh_model.uuids field.

This PR adds BabbleSim test to cover new functionality.

This PR also fixes an issue where an LPN will unnecessarily poll Friend when sending a message over loopback.

PavelVPV committed:
Bluetooth: Mesh: Don't poll friend when sending over loopback

When LPN sends a segmented message to itself, it unnecesseraly polls
Friend. Since Friend doesn't receive this message, the segmented message
transmission will eventually fail (send end callback will return error),
while the message will actually be passed to the access layer.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
Bluetooth: Mesh: Add full virtual addresses support

This commit adds the following features related to virtual addresses
support:
- Allows to store Label UUIDs which virtual addresses collide;
- Allows to decrypt messages encrypted with a virtual address with
collision;
- Allows to publish a message to a specific Label UUID to avoid virtual
addresses collision by adding a pointer to Label UUID to
struct bt_mesh_msg_ctx and struct bt_mesh_model_pub;
- Allows to differentiate Label UUIDs in the model's Subscription List
by storing all subscribed UUIDs in struct bt_mesh_model.uuids field.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bsim: bluetooth: mesh: Test persistent storage of virtual addrs

Test that:
- virtual addresses in subscription list and publication states are
stored and restored correctly
- virtual addresses with collisions are stored and restored correctly

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bsim: bluetooth: mesh: Test tran to virtual addr with collision

Test that transport layer correctly encrypts and decrypts messages to
virtual addresses with collision.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bsim: bluetooth: mesh: Test friendship subs virtual addrs coll

Test that LPN correctly sends Friend Subscription List Add and Remove
messages to Friend when subscribed and unsubscribed to virtual addresses
wirth collision.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Andrewpini reviewed (commented):
LGTM 👍 I have a couple of minor comments

alxelax commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -1412,49 +1414,63 @@ static int mod_sub_va_add(struct bt_mesh_model *model,
+	/* bt_mesh_model_find_uuid() should find a model where both, uuids and groups lists have
+	 * empty entry.
+	 */
+	__ASSERT_NO_MSG(group_entry);
```
The device shouldn't fall into assertion. It should return the error status back.

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/friendship/msg_va_collision.sh:
```diff
@@ -0,0 +1,15 @@
+
+source $(dirname "${BASH_SOURCE[0]}")/../../_mesh_test.sh
+
+# Test receives on group and virtual addresses in the LPN
```
Could you add a bit more description of the test scenario? It is quite a complex sequence.

PavelVPV commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -467,7 +467,7 @@ static void seg_tx_send_unacked(struct seg_tx *tx)
 end:
 	if (IS_ENABLED(CONFIG_BT_MESH_LOW_POWER) &&
-	    bt_mesh_lpn_established()) {
+	    bt_mesh_lpn_established() && !bt_mesh_has_addr(ctx.addr)) {
```
The polling is needed because the transport layer requires ack. If LPN will wait for the next poll timeout, the segmented message transmission will most likely be aborted.

PavelVPV commented on include/zephyr/bluetooth/mesh/access.h:
```diff
@@ -17,10 +17,18 @@
 	{ LISTIFY(_grps, BT_MESH_ADDR_UNASSIGNED_ELT_, (,)) }
+#if CONFIG_BT_MESH_LABEL_COUNT > 0
+#define BT_MESH_MODEL_UUIDS_UNASSIGNED()			\
+	(const uint8_t *[]){ LISTIFY(CONFIG_BT_MESH_LABEL_COUNT, \
```
Then I need to change other cases where the same approach is used. But isn't it clear in the upper layer macro where this one is used?

alxelax replied:
which cases do you mean?

PavelVPV replied:
Ah, apologize. I though I used `.uuid = ` part in models init macros. Did as you suggested.

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1002,6 +1006,69 @@ uint16_t *bt_mesh_model_find_group(struct bt_mesh_model **mod, uint16_t addr)
+	int i;
+
+	for (i = 0; i < CONFIG_BT_MESH_LABEL_COUNT; i++) {
+		if (mod->uuids[i] == uuid) {
```
`bt_mesh_model_find_uuid` is internal and is not intended to be used by customers. The idea is exactly to avoid `memcmp` here and in all other code as it is much more efficient to do this way. 

Allowing customers to use their uuid creates 2 issues. First, it would require them to keep the memory with uuid until the message is sent out to the air. This is because of the case where Friend node re-encrypts own message in Friend Queue targeted to LPN (see `unseg_app_sdu_prepare` in `friend.c`). Second, even if they are able to send the message using their own uuid, for receiving the message the transport layer still needs to know the uuid in advance. And in this case uuid should be added to the stack anyway. Which raises a question: why not add virtual address to the stack in advance and then use the pointer everywhere and avoid `memcmp`?

One more point to the design I think I should emphasize.

For customers now there are 2 places where they have ability to set virtual address as a destination: through `bt_mesh_msg` and `bt_mesh_model_pub` structures. The second one should not really be altered by application and should only be controlled by Configuration Client.

From the spec point of view there is no way to simply add or remove virtual address. Virtual address is only added through subscription or publication states. And when added, it also affects LPN as the node needs send a subscription message to its friend. When the address is removed, all models should be unsubscribed from it, as well as LPN.

I intentionally didn't leave a public API to add/delete/get virtual addresses for app, because the cases above need to be considered in such API, but the benefit is not clear for me. In addition to this, when the access layer receives a virtual address message, it should find a model subscribed to that address. So without the subscribing a model to the address through Configuration Client, usage of virtual address on its own makes less sense. We have a callback that can be used to process any received message (`CONFIG_BT_MESH_ACCESS_LAYER_MSG`). This is I guess they only case where it could potentially be used. I may add getter for the address, if you think it is necessary. But adding and removing should IMO be done through Configuration Client.

alxelax replied:
It is difficult to say what is better for virtual address usage. I didn't see real use cases. Let's leave it `as is` if anybody claims any problem with it then it will be, at least, a direction for improvement. right now there is no reason to reach `perfect` solution.

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1614,20 +1741,39 @@ static int mod_set_pub(struct bt_mesh_model *mod, size_t len_rd,
 	if (err) {
-		LOG_ERR("Failed to set \'model-pub\'");
-		return err;
+		err = bt_mesh_settings_set(read_cb, cb_arg, &pub, sizeof(pub.base));
```
Otherwise the customers will have to unprovision devices before upgrading to the new version, which is, for example, possible through P2P DFU. We already have this backward compatibility in, for example, CDB. Not that I like this, but this creates less pain for those who already is the stack.

alxelax replied:
How about adding it under deprecation macro and then removing it after a couple of releases? 
My concern is that this solves the immediate problem and then becomes the useless ballast that we have to maintain.

PavelVPV replied:
Added a deprecated option that disables the recovery code. By default, the option is disabled and the recovery code is present. This way it won't make a noise when building applications and won't require users to add anything in their project file. They can still remove support for the recovery code by enabling this option. After the deprecation period we will remove this option and the recovery code. Those who enabled this option will need to remove this after the deprecation period. Thanks for the suggestion!

PavelVPV commented on subsys/bluetooth/mesh/access.c:
```diff
@@ -1810,15 +1996,21 @@ static void store_pending_mod_pub(struct bt_mesh_model *mod, bool vnd)
+		}
 
-		err = settings_save_one(path, &pub, sizeof(pub));
+		err = settings_save_one(path, &pub,
```
`base` substructure came from backward compatibility. Even if I remove this and store the whole structure, the code for reading it will stay as is to keep backward compatibility. I can swap around line 1744 and 1742 so that `base` version is always read first as most likely to happen.

PavelVPV commented on subsys/bluetooth/mesh/transport.c:
```diff
@@ -1104,7 +1104,7 @@ static int send_ack(struct bt_mesh_subnet *sub, uint16_t src, uint16_t dst,
 	LOG_DBG("SeqZero 0x%04x Block 0x%08x OBO %u", seq_zero, block, obo);
 
-	if (bt_mesh_lpn_established()) {
+	if (bt_mesh_lpn_established() && !bt_mesh_has_addr(ctx.addr)) {
```
Updated here: https://github.com/zephyrproject-rtos/zephyr/pull/57878/commits/cbda73ef5afcd62e297565d6e58cb30621732667

alxelax reviewed (commented):
LGTM just needs to clarify the final case with backward compatibility. After this, I'll approve PR.

PavelVPV committed:
Bluetooth: Mesh: Add option to disable Label UUIDs recovery

After adding support for virtual addresses with collision (where two
Label UUIDs have the same virtual address), the format of the data in
the persistent storage with the Label UUIDs which a model is subscribed
to or publishes to has been changed. The recovery code is added and the
Label UUIDs will be recovered by picking first Label UUID matching to
the virtual address in the subscription list or model publication. This
options can disable the recovery code and save some flash if the
recovery is not required (e.g. virtual address support wasn't enabled
before this option was added, or the devices were unprovisioned before
upgrading to the version with this option).

Making this option as deprecated to be able to drop support of this
option and remove the recovery code eventually.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>
